pkg/templates/generic_lookup_func.go

```diff
@@ -297,3 +299,109 @@ func onAllowlist(allowlist []ClusterScopedObjectIdentifier, rsrc ClusterScopedOb
 
 	return false
 }
+
+func (t *TemplateResolver) getNodesWithRoleHelper(
+	options *ResolveOptions,
+	templateResult *TemplateResult,
+) func(...string) (
+	map[string]interface{}, error,
+) {
+	return func(name ...string) (
+		map[string]interface{}, error,
+	) {
+		return t.getNodesWithRole(options, templateResult, name...)
+	}
+}
+
+// function getNodesWithRole gets a listing of all nodes with only the
+// role specified.  Any nodes that include other roles in addition
+// to the specified roles are not included.
+func (t *TemplateResolver) getNodesWithRole(
+	options *ResolveOptions,
+	templateResult *TemplateResult,
+	name ...string,
+) (
+	map[string]interface{}, error,
+) {
+	result := []unstructured.Unstructured{}
+
+	if len(name) == 0 || strings.TrimSpace(name[0]) == "" {
+		return nil, fmt.Errorf("%w: name must be specified", ErrInvalidInput)
+	}
+
+	var searchRoles []string
+
+	for _, n := range name {
+		searchRoles = append(searchRoles, fmt.Sprintf("node-role.kubernetes.io/%v", n))
+	}
+
+	nodes, err := t.getOrList(options, templateResult, "v1", "Node", "", "", searchRoles...)
+	if err != nil {
+		return nil, err
+	}
+
+	var nodeList unstructured.UnstructuredList
+
+	nlByte, err := json.Marshal(nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	_ = json.Unmarshal(nlByte, &nodeList)
+
+	// check if the nodes found contain any other role labels
+	for _, node := range nodeList.Items {
+		isMatchedNode := true
+
+		for k := range node.GetLabels() {
+			if strings.HasPrefix(k, "node-role.kubernetes.io") &&
+				!slices.Contains(searchRoles, k) &&
```

Comment:
Is this needed since the label selector is used above to perform server-side filtering?

Reply:
The selector would limit nodes to the role specified, this then checks every label to discard nodes that have the selector and any additional role labels.

It is common to include both an infra role and storage role for ODF as an example. The purpose if the function is to save the user from having to know what additional roles might be used within a cluster.

The check here is looping all labels on the node; ignore any that don't use the standard role notation, and compare those that are a role with the role specified to determine if that node should be included in the set.

Reply:
I see, that makes sense. The template function name doesn't really make that clear. Does renaming the templating function to something like `getNodesWithExactRoles` make sense? An alternative might be to have a `allowOtherRoles=false` argument in the template function.